src/object/cli_shard.c

```diff
@@ -2089,6 +2094,7 @@ obj_shard_query_key_cb(tse_task_t *task, void *data)
 	flags = okqi->okqi_api_flags;
 	opc = opc_get(cb_args->rpc->cr_opc);
 
+	*cb_args->map_ver = obj_reply_map_version_get(rpc);
```

Comment:
Do you mean that *cb_args->map_ver is not initialized to 0 before obj_shard_query_key_cb is called? Why dc_rw_cb does not need this change?

Reply:
oh, map_ver needs to be updated even for failure case, -2007 etc,  otherwise it will fail out in 2124 line, so map_ver can not be updated,  then pool map task will not be triggered in this case.

Reply:
Would it be more natural to place this line around line 2098 (a line number without this patch)? That is, if we get a reply, read the pool map version field in the reply, regardless of the reply status?

Sorry for being slow, but I still don't get why dc_rw_cb does not suffer from this issue. Would you mind giving a hint?

Reply:
And, dc_enumerate_cb? Is obj_shard_query_key_cb the only criminal?

Reply:
>Do you mean that *cb_args->map_ver is not initialized to 0 before obj_shard_query_key_cb is called? Why dc_rw_cb does not need this change?

Hmm, I think you are right about map_ver is not reset to 0 for retry case.  But I will force map_refresh will be triggered in any case if pm_stale == true to make sure there are no missing case for the moment.

And we may refine it later for all RPCs to use this map_ver checking, let's see.

Reply:
We will have to revert the change made to obj_retry_cb when we fix this issue properly. Are you okay with that?

Reply:
hmm, you mean revert the change in this patch?  this is to make sure if obj_comp_cb() decide to refresh the map (pm_stale = true), then map_refresh will do real refresh, regardless what is the current map reply version.  i.e. fix all RPC temporarily.

Reply:
Yes, by "the change" I mean the change made to obj_retry_cb by this patch. It brings in an inconsistency with other callers (e.g., dtx, oid_alloc) of map_refresh, and may cause unnecessary I/Os when the server side does return a pool map version. So before approving this patch, I would like to make you will be okay with we change obj_retry_cb back in the future.

Reply:
Hmm,I am not sure if map version alone can decide if the pool map refresh is needed, especially if the IO includes multiple shards,  what is your thought here?

Reply:
The map_version parameter of dc_pool_map_refresh_task means the pool map version one knows and therefore wants. If that number != 0 and <= the version we have already cached, we know there's no need to do any POOL_TGT_QUERY_MAP. If that number == 0, it means one doesn't know what version she wants and just want to try, in which case we try a random engine only. So, if an I/O gets pool map versions 10, 10, 11 from 3 shards, and if it passes 0 to dc_pool_map_refresh_task, we try one engine. This engine only has version 10. But map_refresh isn't told about 11, so it completes this map_refresh task. Then, unnecessary retry I/O is sent with version 10, leading to another round map_refresh.

Reply:
>So, if an I/O gets pool map versions 10, 10, 11 from 3 shards, and if it passes 0 to dc_pool_map_refresh_task, we try one engine. 

Hmm, the current I/O stack does not support this, it will not remember all shards replied map version, when it go back obj_comp_cb(),  that is the reason I use 0 in object_retry_cb() for now.  

BTW: obj_comp_cb actually compare the reply map version and current version already. 

>This engine only has version 10. But map_refresh isn't told about 11, so it completes this map_refresh task. Then, unnecessary retry I/O is sent with version 10, leading to another round map_refresh.

Hmm, I thought map refresh is for pulling the newest map from server anyway, and the current I/O only tell if the map is stale,  (BTW: I still do not get the point of maintaining multiple map version, known version, dc pool map version etc), and it is the map refresh responsibility to decide where to fetch the pool map.

Reply:
> > So, if an I/O gets pool map versions 10, 10, 11 from 3 shards, and if it passes 0 to dc_pool_map_refresh_task, we try one engine.
> 
> Hmm, the current I/O stack does not support this, it will not remember all shards replied map version, when it go back obj_comp_cb(), that is the reason I use 0 in object_retry_cb() for now.

It sounds like this is exactly the issue to fix.

> BTW: obj_comp_cb actually compare the reply map version and current version already.

It sounds like this comparison makes sense only if map_ver_reply is meaningful (i.e., the issue above is fixed), though I'm not familiar enough with this part yet.

> > This engine only has version 10. But map_refresh isn't told about 11, so it completes this map_refresh task. Then, unnecessary retry I/O is sent with version 10, leading to another round map_refresh.
> 
> Hmm, I thought map refresh is for pulling the newest map from server anyway,

In the previous design where every client process goes to the PS leader, yes; in the current design where client processes are randomly distributed across the pool, only if map_refresh is told what version is wanted. (Any better ideas?)

> and the current I/O only tell if the map is stale,

Just tell map_refresh what the highest map version among shards' replies is.

> (BTW: I still do not get the point of maintaining multiple map version, known version, dc pool map version etc), and it is the map refresh responsibility to decide where to fetch the pool map.

How can I do to clarify this? Shall we schedule a meeting?

Reply:
>It sounds like this comparison makes sense only if map_ver_reply is meaningful (i.e., the issue above is fixed), though I'm not familiar enough with this part yet.

Hmm, I am not sure how to make the map version reply meaningful,  for example, if the IO include 4 shards,  3 shards return map ver 10, the last one return failure,  and the client map version is 10,  but the newest map version is actually 11.  So in this case,  map refresh is not really triggered?

Anyway my point is that only the map version itself may not be enough to decide if the map needs to be refreshed. TBH, I do not see much point to make it more complicate either. I mean if the object_comp_cb decide if the pool map needs to refresh, then refresh. 


>In the previous design where every client process goes to the PS leader, yes; in the current design where client processes are randomly distributed across the pool, only if map_refresh is told what version is wanted. (Any better ideas?)

Oh, I agree all query should not go to leader, but I think the point is that  map refresh should try to get newest pool map independently instead of rely on the current I/O. If the map is queried from the stale engine unfortunately, then I/O will make it retry anyway.  

>How can I do to clarify this? Shall we schedule a meeting?

Sure, but please after thanksgiving. Not easy to Skype for now. Thanks.